ArtOfIllusion/src/artofillusion/LayoutWindow.java

```diff
@@ -549,18 +549,18 @@ public int compare(ModellingTool o1, ModellingTool o2)
     toolsMenu = Translate.menu("tools");
     menubar.add(toolsMenu);
 	
-    toolsMenuItem = new BMenuItem [modellingTools.size()];
-    for (int i = 0; i < modellingTools.size(); i++)
-      {
-        BMenuItem item = new BMenuItem(modellingTools.get(i).getName());
-        toolsMenu.add(item);
-        item.setActionCommand("modellingTool");
-        item.addEventLink(CommandEvent.class, this, "modellingToolCommand");
-        toolsMenuItem[i] = item;
-      }
+    for (ModellingTool tool: modellingTools)
+    {
+      BMenuItem item = new BMenuItem(tool.getName());
+      item.setActionCommand("modellingTool");
+      item.addEventLink(CommandEvent.class, this, "modellingToolCommand");
+      item.getComponent().putClientProperty("tool", tool);
```

Comment:
Do you recall the intent of this line? I'm not sure what it is for.

Reply:
Line 556 sets methid to be called on this menu item invocation
Line 557 stores in menuItem which tool need to be invoked in modellingToolCommand